cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,216 @@
+                                      "part=c/part0.parquet", "part=d/part0.parquet",
+                                      "part=e/part0.parquet", "part=f/part0.parquet",
+                                      "part=g/part0.parquet", "part=h/part0.parquet",
+                                      "part=i/part0.parquet", "part=j/part0.parquet"};
+    for (const auto& file_path : files) {
+      ASSERT_OK_AND_ASSIGN(auto result, file_system->GetFileInfo(file_path));
+      ASSERT_EQ(result.type(), fs::FileType::File);
+    }
+  }
+
+ protected:
+  inline static std::shared_ptr<fs::FileSystem> file_system;
+  inline static std::shared_ptr<Table> table;
+  inline static std::shared_ptr<HivePartitioning> partitioning;
+  inline static std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory;
+  inline static std::shared_ptr<parquet::encryption::KmsConnectionConfig>
+      kms_connection_config;
+};
+
+// This test demonstrates the process of writing a partitioned Parquet file with the same
+// encryption properties applied to each file within the dataset. The encryption
+// properties are determined based on the selected columns. After writing the dataset, the
+// test reads the data back and verifies that it can be successfully decrypted and
+// scanned.
+TEST_F(DatasetEncryptionTest, WriteReadDatasetWithEncryption) {
+  // Create decryption properties.
+  auto decryption_config =
+      std::make_shared<parquet::encryption::DecryptionConfiguration>();
+  auto parquet_decryption_config = std::make_shared<ParquetDecryptionConfig>();
+  parquet_decryption_config->Setup(crypto_factory, kms_connection_config,
+                                   std::move(decryption_config));
+
+  // Set scan options.
+  auto parquet_scan_options = std::make_shared<ParquetFragmentScanOptions>();
+  parquet_scan_options->parquet_decryption_config = std::move(parquet_decryption_config);
+
+  auto file_format = std::make_shared<ParquetFileFormat>();
+  file_format->default_fragment_scan_options = std::move(parquet_scan_options);
+
+  // Get FileInfo objects for all files under the base directory
+  fs::FileSelector selector;
+  selector.base_dir = kBaseDir;
+  selector.recursive = true;
+
+  FileSystemFactoryOptions factory_options;
+  factory_options.partitioning = partitioning;
+  factory_options.partition_base_dir = kBaseDir;
+  ASSERT_OK_AND_ASSIGN(auto dataset_factory,
+                       FileSystemDatasetFactory::Make(file_system, selector, file_format,
+                                                      factory_options));
+
+  // Read dataset into table
+  ASSERT_OK_AND_ASSIGN(auto dataset, dataset_factory->Finish());
+  ASSERT_OK_AND_ASSIGN(auto scanner_builder, dataset->NewScan());
+  ASSERT_OK_AND_ASSIGN(auto scanner, scanner_builder->Finish());
+  ASSERT_OK_AND_ASSIGN(auto read_table, scanner->ToTable());
+
+  // Verify the data was read correctly
+  ASSERT_OK_AND_ASSIGN(auto combined_table, read_table->CombineChunks());
```

Comment:
Let's also validate the table.
```suggestion
  ASSERT_OK_AND_ASSIGN(auto combined_table, read_table->CombineChunks());
  ASSERT_OK(combined_table->ValidateFull());
```